### PR TITLE
ADD the basic interface of explainer

### PR DESCRIPTION
**Pull request by mapleFU**
Signed-off-by: mapleFU <1506118561@qq.com>

### Check List 

Tests 

 - Unit test
 - E2E test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Go code change
 - Has CI related scripts change
 - Has Terraform scripts change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

**Comment by mapleFU** [comment]
The combine logic and `check` is still unimplemented now.

**Comment by mapleFU** [comment]
```
{:type :info, :f :leave, :value "n5", :process :nemesis, :time 98858371130, :index 20112}
```

Maybe we cannot handle this.

**Comment by mahjonp** [comment]
> ```
> {:type :info, :f :leave, :value "n5", :process :nemesis, :time 98858371130, :index 20112}
> ```
> 
> Maybe we cannot handle this.

But nemesis record won't effect the correct of logic

File: pkg/elle/core/depend.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+// RealtimeExplainer is Realtime order explainer
+type RealtimeExplainer struct {
+	historyReference History
+	nextIndex        []int
+}
+
+// TODO: here it needs to get the preEnd and the postStart, it maybe complex and introduce another logic here.
+func (r RealtimeExplainer) ExplainPairData(preEnd, postEnd PathType) ExplainResult {
+	postStart := r.historyReference[r.nextIndex[postEnd.Index]]
```
Comment (mahjonp) [inline pkg/elle/core/depend.go:96]: nextIndex only establishes a mapping from invoke to ok|fail|info from `RealtimeGraph`, we also need bi-direction here.

File: pkg/elle/core/depend.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+// CycleExplainer provides the step-by-step explanation of the relationships between pairs of operations
+type CycleExplainer struct{}
+
+// Explain for the whole scc.
+func (c *CycleExplainer) ExplainCycle(explainer DataExplainer, circle Circle) (Circle, []Step) {
+	var steps []Step
+	for i := 1; i < len(circle.Path); i++ {
+		res := explainer.ExplainPairData(circle.Path[i-1], circle.Path[i])
```
Comment (mahjonp) [inline pkg/elle/core/depend.go:164]: use explainCyclePairData method here?

File: pkg/elle/core/depend.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	for _, v := range bindings {
+		seq = append(seq, fmt.Sprintf(" %s = %s", v.Name, v.Operation.String()))
+	}
+	return strings.Join(seq, "\n")
+}
+
+func explainCycleOps(pairExplainer DataExplainer, bindings []OpBinding, steps []Step) string {
+	var explainitions []string
+	for i := 1; i < len(bindings); i++ {
+		res := pairExplainer.ExplainPairData(bindings[i-1].Operation, bindings[i].Operation)
```
Comment (mahjonp) [inline pkg/elle/core/depend.go:201]: don't need explain again, we can reuse steps here, it's the result of explaination

File: pkg/elle/core/depend.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+type OpBinding struct {
+	Operation Op
+	Name      string
+}
+
+func (c *CycleExplainer) RenderCycleExplanation(explainer DataExplainer, circle Circle, steps []Step) string {
+	var bindings []OpBinding
+	for i, v := range circle.Path {
```
Comment (mahjonp) [inline pkg/elle/core/depend.go:177]: a circle: 2 -> 1 -> 2, so we need bind the last item specially here.

**Pull request merged**